Civi/Core/SettingsBag.php

```diff
@@ -462,4 +479,70 @@ public static function getContributionInvoiceSettingKeys(): array {
     return $convertedKeys;
   }
 
+  /**
+   * Compute a DSN from its component parts or vice versa
+   *
+   * @param string $prefix
+   *   The prefix of the DB setting group - e.g. "civicrm" or "civicrm_uf"
+   *
+   * @return array
+   *   Interpolated values
+   */
+  protected function interpolateDsnSettings(string $prefix): array {
+    $computed = [];
+
+    $dsn = $this->get($prefix . '_db_dsn');
+
+    if ($dsn) {
+      // if dsn is set explicitly, use this as the source of truth.
+      // set the component parts in case anyone wants to AppSettings::get them
+      $urlComponents = \parse_url($dsn);
+
+      if (!$urlComponents) {
+        // couldn't parse the dsn so we dont set the components
+        // (it could be a socket rather than a url)
+        return [];
+      }
+
+      foreach (['user', 'password', 'host', 'port'] as $componentKey) {
+        $settingName = $prefix . '_db_' . $componentKey;
+        $value = $urlComponents[$componentKey] ?? NULL;
+
+        if ($value) {
+          $computed[$settingName] = $value;
+        }
+      }
+
+      // for db name we need to parse the path
+      $settingName = $prefix . '_db_name';
+
+      $urlPath = $urlComponents['path'] ?? '';
+      $dbName = trim($urlPath, '/');
+      if ($dbName) {
+        $computed[$settingName] = $dbName;
+      }
+      return $computed;
+    }
+
+    $componentValues = [];
+
+    foreach (['host', 'name', 'user', 'password', 'port'] as $componentKey) {
+      $value = $this->get($prefix . '_db_' . $componentKey);
+      if (!$value) {
+        // missing a required key to compose the dsn - so give up
+        return [];
+      }
+      $componentValues[$componentKey] = $value;
+    }
+
+    $dsn = "mysql://{$componentValues['user']}:{$componentValues['password']}@{$componentValues['host']}:{$componentValues['port']}/{$componentValues['name']}?new_link=true";
```

Comment:
Good catch!

Do you know if it's a straight url-encode or some specific kind of escaping in this dsn-context?

(There was something with dsn encoding on Joomla recently - scanning this makes me think `url_encode` is prob the surest way forward. @totten / @seamuslee001 may know?)

Reply:
Added `urlencode` and tested with a password containing `@` and `#` and seems to work ok. Wondering if there are any other gotchas here...

Reply:
> (@ufundo) Added urlencode and tested with a password containing @ and # and seems to work ok. Wondering if there are any other gotchas here...

I haven't fully grokked the patchset/dataflow yet, but just low-level comments on this function...

1. I suggest giving a docbock with example outputs. When juggling between `parse_url()`, `DB::parseDSN()`, `urlencode()`, and `urldecode()`, it's very easy to get tripped-up, and it helps to have the declaration of True North. My guess is that the intended output is like:

    ```php
     /**
      * @return array
      *   Ex: [
      *     'civicrm_db_dsn' => 'mysql://my_user%21:pass%23word@host.name/db_name',
      *     'civicrm_db_user' => 'my_user!',
      *     'civicrm_db_password' => 'pass#word',
      *     'civicrm_db_host' => 'host.name',
      *     'civicrm_db_database' => 'db_name',
      *   ]
      */
    ```

    Esp: indicate which return values are supposed to be encoded-vs-decoded.

    Other parts of the patchset declare the setting `civicrm_db_password`, so my guess is that `civicrm_db_password` is supposed to be plaintext (`pass#word` not `pass%23word`).

2. This function appears to have two main code-paths (one starting from a single DSN-string; the other starting from multiple-properties). Make sure that both code-paths produce the same kind of output.

3. If you're looking at `civicrm.settings.php` and the content of `define('CIVICRM_DSN', '...')`, the format is historically muddled. When reading this data, it's probably better to use `DB::parseDSN()` rather than `parse_url()`.

   * `parse_url()` *only* reads RFC-compliant URLs.
    * `DB::parseDSN()` reads RFC-compliant URLs *as well as* legacy (non-compliant) URLs.
    * If `civicrm.settings.php` originates in the past, then `define('CIVICRM_DSN', '....')` may store an encoding that is not RFC-compliant. (A password like `ab#cd` may be written as `'mysql://username:ab#cd@hostname/dbname'` -- where the literal `#` is non-compliant/ambiguous.)
    * If `civicrm.settings.php` originates recently, then `define('CIVICRM_DSN', '....')` stores an RFC-compliant URL.
    * "Past" vs" "recent" ==> I'm not certain exactly where the dividing line is for Civi-Drupal or Civi-WP, but we had a gnarly+destructive issue circa 5.66-5.73 when Civi-Joomla started to use `parse_url(CIVICRM_DSN)`. (https://lab.civicrm.org/dev/core/-/issues/5128) This issue was fixed by using `DB::parseDSN()` instead.
 
4. Additionally, note that `parse_url()` and `DB::parseDSN()` have slightly different return values:

    * `parse_url()` outputs `['pass' => 'pass%23word']`,
    * `DB::parseDSN()` outputs `['password' => 'pass#word']`

    My guess is that that `parseDSN()` is also closer here -- and `parse_url()` needs extra calls to `urldecode()`. (*But that depends on the declared/intended return-format...*)

Reply:
Have switched to `parseDSN` and updated the docbloc now